Remove PageUpdate Listener

ckilb opened the pull request:
- remove listener onPageUpdate
- add listener onSuspend
- remove faulty preparation cache

Commit:
- remove listener onPageUpdate
- add listener onSuspend
- remove faulty preparation cache

ckilb commented:
Hey @liskin 
You were right about the caching functionality I implemented - it couldn't work the way I did it. I don't know what was going on in my head.

This PR will remove that part of the caching functionality again that was faulty: Data preparation will not be cached anymore. 
I still kept the cache for fetching the Pocketbook book ID in place though.

The reason I implemented a cache at all is that page turns are _sometimes_ slow and I suspect this plugin to be one of the reasons. To still have improved performance, I propose to remove the progress update at all when _onPageUpdate_ fires. The way I see it, we don't need it (anymore) with every page turn as the other events provided by KOReader should be sufficient.

To be a bit more on the safer side here, I also added an _onSuspend_ listener.

Page turns seem to to be faster now for me, even though there still seems to be some random factor.

What do you think about these changes?

liskin commented:
> You were right about the caching functionality I implemented - it couldn't work the way I did it. I don't know what was going on in my head.

Sorry I didn't reply to you back then, I was extremely busy the last few weeks before holidays, and then just stepped away from all keyboard-equipped computers for a couple weeks.

Glad you figured it out eventually anyway :-)

> The reason I implemented a cache at all is that page turns are _sometimes_ slow and I suspect this plugin to be one of the reasons. To still have improved performance, I propose to remove the progress update at all when _onPageUpdate_ fires. The way I see it, we don't need it (anymore) with every page turn as the other events provided by KOReader should be sufficient.

The slowness is a bit unexpected, as the sync is scheduled to happen 3 seconds after a page turn, and is cancelled and rescheduled if anything happens to trigger the sync in the meantime. So to experience slow page turns, you'd need to do them exactly 3 seconds after the last one, as the scheduled task is executing…

Anyway, I had exactly the same thought (drop `onPageUpdate`) back when I implemented `onFlushSettings`, `onCloseDocument` and `onEndOfBook`, but I vaguely remember that it didn't work reliably without it. I don't remember the exact scenario (silly me for not elaborating in the commit message back then), but I think it was either switching away from KOReader using the Home button, or switching away from KOReader by holding the Home button and then interacting with PocketBook's Task Manager. I think in one or both of these scenarios KOReader races with PocketBook's home screen, and sometimes the home screen would read the database before KOReader had a chance to update it. As a result, it wouldn't show the most recently read book first, or the page position would be wrong.

Since I didn't experience any slowness with this plugin, ever, I decided not to dig much deeper and simply left the `onPageUpdate` hook in place.

Holidays are over now, but I'll be travelling over the next 2 days, so I won't be able to test your changes just yet. Perhaps you'll be able to reproduce the issue I described, and then we can brainstorm what to do about it. Otherwise, feel free to ping me at the end of the week and I'll dig into it.

liskin commented on pocketbooksync.koplugin/main.lua:
```diff
@@ -54,52 +54,46 @@ function PocketbookSync:prepareSync()
 
-    local cacheKey = self.ui.document.file
-
-    if not preparedData[cacheKey] then
```
There's still remnants of `preparedData` elsewhere in the file. Perhaps it would've been wiser to simply revert f5f7d3879a45d781a94e09fd327b24e66da10db2 first, and then implement the other changes in a separate commit?

ckilb replied:
Removed the remnants now. I like to have scheduled sync to be removed. Please see my other comment for motivation.

ckilb commented:
Thanks for your elaborate answer, @liskin  👍

Regarding slowness:
When I turn pages in the native app, it starts turning almost immediately. Then the bottle neck is the Eink screen to actually draw the new content. Same for Kindle Paperwhite, which I also used to compare page turn speed.
In KOReader, it often takes a while - maybe roughly 0.3 seconds - before the screen starts redrawing.

It would be interesting too know, if it's not slow for you because you don't consider ~0.3 seconds delay as slow OR because it's really faster on your device. My device is Inkpad Color 2 btw., maybe yours has a faster CPU?

This slowness is randomly and I can't tell for sure what's the reason. What I experienced:
- If I start reading again after a while, it's almost always slow [maybe because some Pocketbook/KOReader/LuaJIT whatever is warming up? Didn't find any suspicious in the list of top processes though]
- If it's slow and I restart KOReader ("Exit" => "Restart KOReader") it's _sometimes_ faster (like, maybe 0.1 seconds before redrawing) [might be a memory issue/leak/anomaly in KOReader]?

After removing the _onPageUpdate_ listener, still, after using KOReader again after a break , it's slow again. But, it seems to always get faster now after a few minutes or page turns. This didn't seem to happen before my changes - but I can't tell for sure, because it's hard to debug as it's such a random thing.

After removing onPageUpdate I tested the following scenarios:
- Pressing Home button -> works
- Long pressing Home button, then click "Main Menu" -> works
- Long pressing Power button to shut down device, then start it again with "Personalize" -> "Open at Startup" ->"Last Opened Book", then pressing Home button -> works
- Even long pressing Power button to shut down device, then start it again with "Personalize" -> "Open at Startup" -> "Main Menu" -> works

Then, because I want the Home button in KOReader go the KOReader's file browser instead, I remapped it to "None" in "Personalize" -> "Key Mapping". Additionally I remapped the Power button to "Flip task" to switch between main menu and KOReader. Even with flipping tasks it works for me.

Maybe (hopefully?) it works for me because I added _onShutdown_? The only reason I could imagine it would not work is if the battery goes down, but I think we can ignore that.

If you could test again on your device, that would be great :+1:

=====

The reason to remove the scheduling is because firstly I never knew and still don't know if it actually has any performance improvement (I initially just added it because it felt the right thing to do) and secondly it messes around with one of my other plugins (Dictionary Mode).

If dictionary mode is enabled, a single short tap on a word will open the default dictionary modal. There are 3 ways to close the dictionary modal: first one is tapping the "Close" button, second one is tapping the "X" button, and third one is to just tap outside of the modal. Usually I do the third one because I don't need to target a specific area.

With scheduling enabled, _sometimes_ (= randomly again) tapping outside of the modal did just have no effect. I had to tap again to make it work. Once I removed scheduling, it _always_ worked. 
As the Dictionary Mode plugin doesn't do anything special here - it basically just opens the default dictionary modal - I think the chance is quite high that scheduling might mess with other features or plugins, too. Therefore, because I didn't see any benefit in scheduling, I removed it.

liskin commented:
Sorry I kinda forgot about this one. Still very busy these days though. I don't recall experiencing any slowness like the one you describe, and my device is quite old - InkPad 3, but then I have quite aggressive power saving settings - power off after an hour of sleep or something. So I rarely resume KOReader after a long wait. If I hadn't been reading for a few hours, the device is powered off.

The only other thing that springs to mind is that setting in gesture manager that lets you configure what is considered a double tap. It defaults to 0.3s, and I think that's how long it waits after a single tap to know it was not a double tap. I don't use double taps, so I disabled this.

I still intend to test on my device but busy busy lately. :-(

liskin commented on pocketbooksync.koplugin/main.lua:
```diff
@@ -39,7 +38,6 @@ local PocketbookSync = WidgetContainer:extend{
 
 function PocketbookSync:clearCache()
```
This is never being called btw. The only way to clear the cache is to restart KOReader.
(Which is probably fine as long as people don't do a chain of renames such as A → B, C → A — then opening the new A will update the old A's library entry. There probably isn't a nice way to handle this.)

liskin commented:
Okay, I just spent a couple minutes testing it and the removal of `onPageUpdate` seems to be fine indeed. The home screen looks correct no matter what weird stuff I try :-)
